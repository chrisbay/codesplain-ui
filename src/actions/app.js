import axios from 'axios';
import cookie from 'react-cookie';

export const CLEAR_UNSAVED_CHANGES = 'CLEAR_UNSAVED_CHANGES';
export const EDIT_ANNOTATION = 'EDIT_ANNOTATION';
export const PARSE_SNIPPET = 'PARSE_SNIPPET';
export const RESTORE_STATE = 'RESTORE_STATE';
export const RESET_RULE_FILTERS = 'RESET_RULE_FILTERS';
export const SAVE_ANNOTATION = 'SAVE_ANNOTATION';
export const SET_AST = 'SET_AST';
export const SET_RULE_FILTERS = 'SET_RULE_FILTERS';
export const SET_SNIPPET_CONTENTS = 'SET_SNIPPET_CONTENTS';
export const SET_SNIPPET_LANGUAGE = 'SET_SNIPPET_LANGUAGE';
export const SET_SNIPPET_TITLE = 'SET_SNIPPET_TITLE';
export const TOGGLE_EDITING_STATE = 'TOGGLE_EDITING_STATE';
export const SAVE_STATE = 'SAVE_STATE';
export const SAVE_STATE_STARTED = 'SAVE_STATE_STARTED';
export const SAVE_STATE_SUCCEEDED = 'SAVE_STATE_SUCCEEDED';
export const SAVE_STATE_FAILED = 'SAVE_STATE_FAILED';
export const SELECT_ALL_FILTERS = 'SELECT_ALL_FILTERS';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const UPDATE_USER_SNIPPETS_STARTED = 'UPDATE_USER_SNIPPETS_STARTED';
export const UPDATE_USER_SNIPPETS_SUCCEEDED = 'UPDATE_USER_SNIPPETS_SUCCEEDED';
export const UPDATE_USER_SNIPPETS_FAILED = 'UPDATE_USER_SNIPPETS_FAILED';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';

const API_URL = process.env.REACT_APP_API_URL;

export const setSnippetContents = (snippet) => ({
  type: SET_SNIPPET_CONTENTS,
  payload: snippet,
});

export const setRuleFilters = (filters) => ({
  type: SET_RULE_FILTERS,
  payload: filters,
});

export const selectAllFilters = () => ({
  type: SELECT_ALL_FILTERS,
});

export const resetFilters = () => ({
  type: RESET_RULE_FILTERS,
});

export const setAST = (AST) => ({
  type: SET_AST,
  payload: AST,
});

export const toggleEditState = () => ({
  type: TOGGLE_EDITING_STATE,
});

export const saveAnnotation = (annotationData) => ({
  type: SAVE_ANNOTATION,
  payload: annotationData,
});

export const restoreState = (savedState) => ({
  type: RESTORE_STATE,
  payload: savedState,
});

export const setSnippetTitle = (snippetTitle) => ({
  type: SET_SNIPPET_TITLE,
  payload: snippetTitle
})

export const setSnippetLanguage = (snippetLanguage) => ({
  type: SET_SNIPPET_LANGUAGE,
  payload: snippetLanguage
});

export const parseSnippet = (snippet) => ({
  type: PARSE_SNIPPET,
  meta: {
    WebWorker: true,
  },
  payload: {
    snippet,
  },
});

export const clearUnsavedChanges = () => ({
  type: CLEAR_UNSAVED_CHANGES,
})

export const saveStateStarted = () => ({
  type: SAVE_STATE_STARTED,
});

export const saveStateSucceeded = () => ({
  type: SAVE_STATE_SUCCEEDED,
});

export const saveStateFailed = () => ({
  type: SAVE_STATE_FAILED,
});

export const setUserSnippets = (snippetMeta) => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const updateUserSnippetsStarted = () => ({
  type: UPDATE_USER_SNIPPETS_STARTED,
});

export const updateUserSnippetsSucceeded = () => ({
  type: UPDATE_USER_SNIPPETS_SUCCEEDED,
});

export const updateUserSnippetsFailed = () => ({
  type: UPDATE_USER_SNIPPETS_FAILED,
});

export const updateUserSnippets = () => {
  return (dispatch) => {
    // Load requisite cookies, return Promise if they aren't present
    const token    = cookie.load('token');
    const username = cookie.load('username');
    if (!token || !username) {
      return Promise.resolve();
    }

    // Fetch the user's snippet meta data and save it
    const headers = {
      Accept: 'application/json',
      Authorization: `token ${token}`,
    };
    dispatch(updateUserSnippetsStarted());
    return axios.get(`${API_URL}/users/${username}/snippets`, { headers })
      .then(
        res => {
          dispatch(setUserSnippets(res));
          dispath(updateUserSnippetsSucceeded());
        },
        err => dispatch(updateUserSnippetsFailed)
      );
  };
};

export const saveState = (id) => {
  return (dispatch, getState) => {
    // If the snippet has not been saved, then saving an annotation shouldn't
    // trigger a POST request to save the snippet (and its annotations)
    const username = cookie.load('username');
    if (id === undefined || !username) {
      // Return a Promise object to make this action "thenable"
      return Promise.resolve();
    }
    dispatch(saveStateStarted());
    // Get the app state and stringify it
    const stateString = JSON.stringify(getState().app);
    return axios.post(`${API_URL}/users/${username}/snippets/${id}`, { json: stateString, })
      .then((res) => {
        dispatch(saveStateSucceeded());
      }, (err) => {
        dispatch(saveStateFailed());
      });
  };
};
