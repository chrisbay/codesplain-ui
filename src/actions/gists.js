import {
  parseSnippet,
  setSnippetContents,
} from './app';
import {
  fetchGist as makeFetchGistRequest,
  fetchGists as makeFetchGistsRequest,
} from '../util/requests';

export const SET_GISTS = 'SET_GISTS';

export const setGists = gists => ({
  type: SET_GISTS,
  payload: gists,
});

export const fetchGist = url => dispatch => (
  makeFetchGistRequest(url)
    .then((contents) => {
      dispatch(setSnippetContents(contents));
      dispatch(parseSnippet(contents));
    })
);

export const fetchGists = () => (dispatch, getState) => {
  const { token } = getState().user;
  return makeFetchGistsRequest(token)
    .then((gists) => { dispatch(setGists(gists)); });
};
