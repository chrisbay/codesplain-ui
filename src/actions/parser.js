export const SET_PARSER_STATE = 'SET_PARSER_STATE';
export const LOAD_PARSER = 'LOAD_PARSER'
export const LOADING = 'LOADING'
export const LOADED = 'LOADED'

export const setParserState = (state) => ({
  type: SET_PARSER_STATE,
  payload: state
})

export const loadParser = (parser) => ({
  type: LOAD_PARSER,
  payload: parser
})
