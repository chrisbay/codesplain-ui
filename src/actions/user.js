import axios from 'axios';
import _ from 'lodash';

import { makeSaveEndpointUrl, sanitizeSnippetList } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const ADD_ORG = 'ADD_ORG';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const SET_SNIPPET_LISTS = 'SET_SNIPPET_LISTS';
export const SWITCH_ORG = 'SWITCH_ORG';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';
export const UPDATE_USER_SNIPPETS_FAILED = 'UPDATE_USER_SNIPPETS_FAILED';
export const UPDATE_USER_SNIPPETS_STARTED = 'UPDATE_USER_SNIPPETS_STARTED';
export const UPDATE_USER_SNIPPETS_SUCCEEDED = 'UPDATE_USER_SNIPPETS_SUCCEEDED';

export const addOrg = org => ({
  type: ADD_ORG,
  payload: org,
});

export const addOrganizations = organizations => ({
  type: ADD_ORGANIZATIONS,
  payload: organizations,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const setSnippetLists = snippetLists => ({
  type: SET_SNIPPET_LISTS,
  payload: snippetLists,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const switchOrg = org => ({
  type: SWITCH_ORG,
  payload: org,
});

export const updateUserSnippetsStarted = () => ({
  type: UPDATE_USER_SNIPPETS_STARTED,
});

export const updateUserSnippetsSucceeded = () => ({
  type: UPDATE_USER_SNIPPETS_SUCCEEDED,
});

export const updateUserSnippetsFailed = () => ({
  type: UPDATE_USER_SNIPPETS_FAILED,
});

export const updateUserSnippets = () => (dispatch, getState) => {
  // Load requisite cookies, return Promise if they aren't present
  const {
    user: {
      token,
      username,
    },
  } = getState();
  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  dispatch(updateUserSnippetsStarted());
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then((res) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(res.data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(res.data));
      dispatch(updateUserSnippetsSucceeded());
    })
    .catch(() => {
      dispatch(updateUserSnippetsFailed());
    });
};

export const fetchSnippetLists = () => (dispatch, getState) => {
  const {
    token,
    orgs,
  } = getState().user;
  const headers = {
    Authorization: `token ${token}`,
  };
  const url = `${API_URL}/users?users=${orgs.join(',')}`;
  // Remove the username from the org list
  return axios({
    method: 'GET',
    url,
    headers,
  })
    .then(({ data: snippetLists }) => {
      const sanitizedSnippetLists = _.mapValues(snippetLists, sanitizeSnippetList);
      dispatch(setSnippetLists(sanitizedSnippetLists));
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const restoreUserCredentials = (token, username) => ({
  type: RESTORE_USER_CREDENTIALS,
  payload: {
    token,
    username,
  },
});
