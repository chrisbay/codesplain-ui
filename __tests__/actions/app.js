import moxios from 'moxios';

import * as actions from '../../src/actions/app';
import generateMockStore from '../../src/testUtils/mockStore';

describe('Actions: App', () => {
  describe('SET_SNIPPET_KEY', () => {
    it('creates correct action object', () => {
      const key = 'szechuan sauce';
      expect(actions.setSnippetKey(key)).toMatchSnapshot();
    });
  });
  describe('SET_SNIPPET_CONTENTS', () => {
    it('creates an action to set the snippet contents', () => {
      const snippet = 'Show me what you got';
      expect(actions.setSnippetContents(snippet)).toMatchSnapshot();
    });
  });
  describe('SET_RULE_FILTERS', () => {
    it('creates an action to set the rule filters', () => {
      const filters = {
        for_stmt: {
          prettyTokenName: 'For Loops',
          count: 1,
          selected: false,
        },
        atom: {
          prettyTokenName: 'Atoms',
          count: 21,
          selected: true,
        },
      };
      expect(actions.setRuleFilters(filters)).toMatchSnapshot();
    });
  });
  describe('RESET_RULE_FILTERS', () => {
    it('matches snapshot', () => {
      expect(actions.resetFilters()).toMatchSnapshot();
    });
  });
  describe('SELECT_ALL_FILTERS', () => {
    it('matches snapshot', () => {
      expect(actions.selectAllFilters()).toMatchSnapshot();
    });
  });
  describe('RESET_STATE', () => {
    it('matches snapshot', () => {
      expect(actions.resetState()).toMatchSnapshot();
    });
  });
  describe('SET_AST', () => {
    it('creates an action to set the abstract syntax tree', () => {
      const AST = {
        type: 'file_input',
        begin: 0,
        end: 1,
        children: [
          'NEWLINE',
          'INDENT',
          'NEWLINE',
          'DEDENT',
        ],
      };
      expect(actions.setAST(AST)).toMatchSnapshot();
    });
  });
  describe('TOGGLE_EDITING_STATE', () => {
    it('creates an action that toggles the editing state', () => {
      expect(actions.toggleEditState()).toMatchSnapshot();
    });
  });
  describe('SAVE_ANNOTATION', () => {
    it('creates an action that adds an annotation', () => {
      const annotationData = {
        annotation: 'You pass butter.',
        lineNumber: 1,
        lineText: 'What is my purpose?',
      };
      expect(actions.saveAnnotation(annotationData)).toMatchSnapshot();
    });
  });
  describe('RESTORE_STATE', () => {
    it('creates an action that restores state', () => {
      const savedState = {
        snippet: '\t',
        snippetTitle: '',
        annotations: {},
        AST: {
          type: 'file_input',
          begin: 0,
          end: 1,
          children: [
            'NEWLINE',
            'INDENT',
            'NEWLINE',
            'DEDENT',
          ],
        },
        filters: {},
        readOnly: false,
      };
      expect(actions.restoreState(savedState)).toMatchSnapshot();
    });
  });
  describe('SET_SNIPPET_TITLE', () => {
    it('creates an action to set the snippet title', () => {
      const snippetTitle = 'Get Schwifty';
      const expected = {
        type: actions.SET_SNIPPET_TITLE,
        payload: snippetTitle,
      };
      expect(actions.setSnippetTitle(snippetTitle)).toEqual(expected);
    });
  });
  describe('SET_SNIPPET_LANGUAGE', () => {
    it('creates an action to set the snippet language', () => {
      const snippetLanguage = 'Gromflomite';
      expect(actions.setSnippetLanguage(snippetLanguage)).toMatchSnapshot();
    });
  });
  describe('PARSE_SNIPPET', () => {
    it('creates an action to parse the snippet with a web worker', () => {
      const snippet = 'Grasssss tastes bad';
      expect(actions.parseSnippet(snippet)).toMatchSnapshot();
    });
  });
  describe('SAVE_SUCCEEDED', () => {
    it('creates the correct action object', () => {
      const expected = {
        type: actions.SAVE_SUCCEEDED,
      };
      expect(actions.saveSucceeded()).toEqual(expected);
    });
  });
  describe('async actions', () => {
    beforeEach(() => {
      moxios.install();
    });
    afterEach(() => {
      moxios.uninstall();
    });
    describe('saveNew()', () => {
      it('dispatches SAVE_SUCCEEDED if saved', () => {
        const key = 'gazorpazorp';
        const org = 'rick-sanchez';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: { key, status: '200' },
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '' },
        });
        return store.dispatch(actions.saveNew(org))
          .then((returnVal) => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
            expect(returnVal).toEqual(key);
          });
      });
      it('dispatches SAVE_FAILED if save failed', () => {
        const org = 'rick-sanchez';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '' },
        });
        return store.dispatch(actions.saveNew(org))
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
    describe('deleteSnippet()', () => {
      it('dispatches the correct actions if the action succeeded', () => {
        const snippetKey = 'snippetKey';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
          });
        });
        const store = generateMockStore({
          user: {
            token: '',
            selectedOrg: '',
          },
        });
        return store.dispatch(actions.deleteSnippet(snippetKey))
          .then(() => {
            expect(store.getActions()).toMatchSnapshot();
          });
      });
      it('dispatches the correct actions if the action fails', () => {
        const snippetKey = 'snippetKey';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          user: {
            token: '',
            selectedOrg: '',
          },
        });
        return store.dispatch(actions.deleteSnippet(snippetKey))
          .catch(() => {
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
    describe('saveExisting()', () => {
      it('dispatches SAVE_SUCCEEDED if saved', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: { status: '200' },
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '', username: '' },
        });
        return store.dispatch(actions.saveExisting())
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
      it('dispatches SAVE_FAILED if save failed', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '', username: '' },
        });
        return store.dispatch(actions.saveExisting())
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
  });
});
