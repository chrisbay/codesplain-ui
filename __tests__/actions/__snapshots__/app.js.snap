exports[`Actions: App PARSE_SNIPPET creates an action to parse the snippet with a web worker 1`] = `
Object {
  "meta": Object {
    "WebWorker": true,
  },
  "payload": Object {
    "snippet": "Grasssss tastes bad",
  },
  "type": "PARSE_SNIPPET",
}
`;

exports[`Actions: App RESET_RULE_FILTERS matches snapshot 1`] = `
Object {
  "type": "RESET_RULE_FILTERS",
}
`;

exports[`Actions: App RESET_STATE matches snapshot 1`] = `
Object {
  "type": "RESET_STATE",
}
`;

exports[`Actions: App RESTORE_STATE creates an action that restores state 1`] = `
Object {
  "payload": Object {
    "AST": Object {
      "begin": 0,
      "children": Array [
        "NEWLINE",
        "INDENT",
        "NEWLINE",
        "DEDENT",
      ],
      "end": 1,
      "type": "file_input",
    },
    "annotations": Object {},
    "filters": Object {},
    "readOnly": false,
    "snippet": "	",
    "snippetTitle": "",
  },
  "type": "RESTORE_STATE",
}
`;

exports[`Actions: App SAVE_ANNOTATION creates an action that adds an annotation 1`] = `
Object {
  "payload": Object {
    "annotation": "You pass butter.",
    "lineNumber": 1,
    "lineText": "What is my purpose?",
  },
  "type": "SAVE_ANNOTATION",
}
`;

exports[`Actions: App SELECT_ALL_FILTERS matches snapshot 1`] = `
Object {
  "type": "SELECT_ALL_FILTERS",
}
`;

exports[`Actions: App SET_AST creates an action to set the abstract syntax tree 1`] = `
Object {
  "payload": Object {
    "begin": 0,
    "children": Array [
      "NEWLINE",
      "INDENT",
      "NEWLINE",
      "DEDENT",
    ],
    "end": 1,
    "type": "file_input",
  },
  "type": "SET_AST",
}
`;

exports[`Actions: App SET_RULE_FILTERS creates an action to set the rule filters 1`] = `
Object {
  "payload": Object {
    "atom": Object {
      "count": 21,
      "prettyTokenName": "Atoms",
      "selected": true,
    },
    "for_stmt": Object {
      "count": 1,
      "prettyTokenName": "For Loops",
      "selected": false,
    },
  },
  "type": "SET_RULE_FILTERS",
}
`;

exports[`Actions: App SET_SNIPPET_CONTENTS creates an action to set the snippet contents 1`] = `
Object {
  "payload": "Show me what you got",
  "type": "SET_SNIPPET_CONTENTS",
}
`;

exports[`Actions: App SET_SNIPPET_KEY creates correct action object 1`] = `
Object {
  "payload": "szechuan sauce",
  "type": "SET_SNIPPET_KEY",
}
`;

exports[`Actions: App SET_SNIPPET_LANGUAGE creates an action to set the snippet language 1`] = `
Object {
  "payload": "Gromflomite",
  "type": "SET_SNIPPET_LANGUAGE",
}
`;

exports[`Actions: App TOGGLE_EDITING_STATE creates an action that toggles the editing state 1`] = `
Object {
  "type": "TOGGLE_EDITING_STATE",
}
`;

exports[`Actions: App async actions deleteSnippet() dispatches the correct actions if the action fails 1`] = `
Array [
  Object {
    "payload": "Deleting...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to delete snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions deleteSnippet() dispatches the correct actions if the action succeeded 1`] = `
Array [
  Object {
    "payload": "Deleting...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Snippet Deleted!",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveExisting() dispatches SAVE_FAILED if save failed 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to save snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveExisting() dispatches SAVE_SUCCEEDED if saved 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Codesplaination Saved!",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "SAVE_SUCCEEDED",
  },
]
`;

exports[`Actions: App async actions saveNew() dispatches SAVE_FAILED if save failed 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to save snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveNew() dispatches SAVE_SUCCEEDED if saved 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Codesplaination Saved!",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "SAVE_SUCCEEDED",
  },
]
`;
